test/unit/org/apache/cassandra/index/sai/functional/FailureTest.java

```diff
@@ -0,0 +1,129 @@
+
+        execute("INSERT INTO %s (id1, v1) VALUES ('1', '1')");
+        execute("INSERT INTO %s (id1, v1) VALUES ('2', '2')");
+        flush();
+
+        assertEquals(1, execute("SELECT id1 FROM %s WHERE v1='1'").size());
+
+        verifyIndexFiles(literalIndexContext, 1, 1, 1);
+        verifySSTableIndexes(literalIndexContext.getIndexName(), 1, 1);
+
+        execute("INSERT INTO %s (id1, v1) VALUES ('3', '3')");
+
+        Injection ssTableContextCreationFailure = newFailureOnEntry("context_failure_on_flush", SSTableContext.class, "<init>", RuntimeException.class);
+        Injections.inject(ssTableContextCreationFailure);
+
+        flush();
+
+        // Verify that, while the node is still operational, the index is not.
+        Assertions.assertThatThrownBy(() -> execute("SELECT * FROM %s WHERE v1='1'"))
+                  .isInstanceOf(IndexNotAvailableException.class);
+
+        ssTableContextCreationFailure.disable();
+
+        // Now verify that a restart actually repairs the index...
+        simulateNodeRestart();
+
+        verifyIndexFiles(literalIndexContext, 2);
+        verifySSTableIndexes(literalIndexContext.getIndexName(), 2, 2);
+
+        assertEquals(1, execute("SELECT id1 FROM %s WHERE v1='1'").size());
+    }
+
+    @Test
+    public void shouldMakeIndexNonQueryableOnSSTableContextFailureDuringCompaction() throws Throwable
+    {
+        createTable(CREATE_TABLE_TEMPLATE);
+        IndexContext literalIndexContext = createIndexContext(createIndex(String.format(CREATE_INDEX_TEMPLATE, "v1")), Int32Type.instance);
+
+        execute("INSERT INTO %s (id1, v1) VALUES ('1', '1')");
+        flush();
+
+        execute("INSERT INTO %s (id1, v1) VALUES ('2', '2')");
+        flush();
+
+        assertEquals(1, execute("SELECT id1 FROM %s WHERE v1='1'").size());
+
+        verifyIndexFiles(literalIndexContext, 2, 2, 2);
+        verifySSTableIndexes(literalIndexContext.getIndexName(), 2, 2);
+
+        Injection ssTableContextCreationFailure = newFailureOnEntry("context_failure_on_compaction", SSTableContext.class, "<init>", RuntimeException.class);
+        Injections.inject(ssTableContextCreationFailure);
+
+        compact();
+
+        // Verify that the index is not available.
+        Assertions.assertThatThrownBy(() -> execute("SELECT * FROM %s WHERE v1='1'"))
+                  .isInstanceOf(IndexNotAvailableException.class);
+    }
+
+    @Test
+    public void shouldMakeIndexNonQueryableOnSSTableContextFailureDuringCreation() throws Throwable
+    {
+        createTable(CREATE_TABLE_TEMPLATE);
+
+        execute("INSERT INTO %s (id1, v1) VALUES ('1', '1')");
+        execute("INSERT INTO %s (id1, v1) VALUES ('2', '2')");
+
+        // We need to reference SSTableContext first or the failure injection fails
+        // because byteman can't find the class.
+        SSTableContext.class.getName();
```

Comment:
Do we need this reference? The test seems to pass if we remove it.

Reply:
We don't need it because the code has changed underneath. It was needed in a previous version.